staging/src/k8s.io/client-go/util/workqueue/delaying_queue.go

```diff
@@ -205,65 +227,110 @@ func (q *delayingType) waitingLoop() {
 			return
 		}
 
-		now := q.clock.Now()
-
-		// Add ready entries
-		for waitingForQueue.Len() > 0 {
-			entry := waitingForQueue.Peek().(*waitFor)
-			if entry.readyAt.After(now) {
-				break
-			}
-
-			entry = heap.Pop(waitingForQueue).(*waitFor)
-			q.Add(entry.data)
-			delete(waitingEntryByData, entry.data)
-		}
-
-		// Set up a wait for the first item's readyAt (if one exists)
-		nextReadyAt := never
-		if waitingForQueue.Len() > 0 {
-			if nextReadyAtTimer != nil {
-				nextReadyAtTimer.Stop()
-			}
-			entry := waitingForQueue.Peek().(*waitFor)
-			nextReadyAtTimer = q.clock.NewTimer(entry.readyAt.Sub(now))
-			nextReadyAt = nextReadyAtTimer.C()
-		}
-
 		select {
 		case <-q.stopCh:
 			return
 
 		case <-q.heartbeat.C():
-			// continue the loop, which will add ready items
+			q.processWaitingItems(waitingForQueue, waitingEntryByData)
+			if newTimer := q.calculateNextReadyTimer(waitingForQueue, nextReadyAtTimer); newTimer != nil {
+				nextReadyAt = newTimer
+			}
 
 		case <-nextReadyAt:
-			// continue the loop, which will add ready items
+			q.processWaitingItems(waitingForQueue, waitingEntryByData)
+			if newTimer := q.calculateNextReadyTimer(waitingForQueue, nextReadyAtTimer); newTimer != nil {
+				nextReadyAt = newTimer
+			}
 
 		case waitEntry := <-q.waitingForAddCh:
-			if waitEntry.readyAt.After(q.clock.Now()) {
-				insert(waitingForQueue, waitingEntryByData, waitEntry)
-			} else {
-				q.Add(waitEntry.data)
-			}
+			q.processArrivingEntry(waitingForQueue, waitingEntryByData, waitEntry)
 
 			drained := false
 			for !drained {
 				select {
 				case waitEntry := <-q.waitingForAddCh:
-					if waitEntry.readyAt.After(q.clock.Now()) {
-						insert(waitingForQueue, waitingEntryByData, waitEntry)
-					} else {
-						q.Add(waitEntry.data)
-					}
+					q.processArrivingEntry(waitingForQueue, waitingEntryByData, waitEntry)
 				default:
 					drained = true
 				}
 			}
+			if newTimer := q.calculateNextReadyTimer(waitingForQueue, nextReadyAtTimer); newTimer != nil {
+				nextReadyAt = newTimer
+			}
+
+		case request := <-q.inspect:
+			// allow inspection of the waiting queue for testing purposes
+			q.processInspectRequest(waitingEntryByData, request)
```

Comment:
I'm usually not a fan of helping out test code with production code like this, I'll try and think if there's another way to do it.